### PR TITLE
Be resilient to asset-cache mutating while retrieving assets from the host

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
The asset cache serves more as a weak cache of data, storing information on the server in the hopes it will be useful at future points in time.  It contains a GC thread that wakes up every minute or so and flushes data it thinks isn't needed anymore.  Due to that, access to the data from the cache itself is fundamentally racey and should not be trusted.

### CODE REVIEW

File: src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs [hunk 1]
```diff
@@ -29,39 +29,34 @@ internal sealed partial class AssetProvider(Checksum solutionChecksum, SolutionA
     private readonly SolutionAssetCache _assetCache = assetCache;
     private readonly IAssetSource _assetSource = assetSource;
 
-    private T GetRequiredAsset<T>(Checksum checksum)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs:32]: this operation si fundamentally not supported.  there is no way, ever, to ensure that a checksum you're asking for is actually in the asset provider (unles sit was pinned.  but that's a vanishingly small number of cases).

File: src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs [hunk 2]
```diff
@@ -136,24 +131,60 @@ public async ValueTask SynchronizeProjectAssetsAsync(ProjectStateChecksums proje
             missingChecksumsCount = 0;
             foreach (var checksum in checksums)
             {
-                if (!_assetCache.TryGetAsset<object>(checksum, out _))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs:139]: view with whitespace off.

File: src/Workspaces/Remote/ServiceHub/Host/ChecksumSynchronizer.cs [hunk 3]
```diff
@@ -20,17 +20,22 @@ internal sealed partial class AssetProvider
...
             {
-                await _assetProvider.SynchronizeAssetsAsync(assetHint, checksums, cancellationToken).ConfigureAwait(false);
+                await _assetProvider.SynchronizeAssetsAsync(assetHint, checksums, results, cancellationToken).ConfigureAwait(false);
             }
         }
 
         public async ValueTask SynchronizeSolutionAssetsAsync(Checksum solutionChecksum, CancellationToken cancellationToken)
         {
             SolutionStateChecksums solutionChecksumObject;
+
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Host/ChecksumSynchronizer.cs:38]:
```suggestion
```

File: src/Workspaces/Remote/ServiceHub/Host/ChecksumSynchronizer.cs [hunk 4]
```diff
@@ -47,14 +51,17 @@ public async ValueTask SynchronizeSolutionAssetsAsync(Checksum solutionChecksum,
 
                     solutionChecksumObject.AddAllTo(checksums);
                     checksums.Remove(solutionChecksumObject.Checksum);
-                    await _assetProvider.SynchronizeAssetsAsync(assetHint: AssetHint.None, checksums, cancellationToken).ConfigureAwait(false);
+                    await _assetProvider.SynchronizeAssetsAsync(assetHint: AssetHint.None, checksums, results: null, cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Host/ChecksumSynchronizer.cs:54]: so we could pass a dictionary here that we then read in below.  but i didn't feel like it was worth it.  the data will normally always stay in the asset provider.  and the cahnge below to use GetAssetAsync will normally find it, or *rarely* then go and reretrieve it in the rare unlucky case.  that saves a potentially large dictionary alloc here (even if pooled).

File: src/Workspaces/Remote/ServiceHub/Host/ChecksumSynchronizer.cs [hunk 5]
```diff
@@ -82,24 +89,28 @@ private async ValueTask SynchronizeProjectAssets_NoLockAsync(ProjectStateChecksu
...
-            CollectChecksumChildren(this, projectChecksum.AdditionalDocuments.Checksums);
-            CollectChecksumChildren(this, projectChecksum.AnalyzerConfigDocuments.Checksums);
+            // Then synchronize the info about all the documents within.
+            await CollectChecksumChildrenAsync(this, projectChecksum.Documents.Checksums).ConfigureAwait(false);
+            await CollectChecksumChildrenAsync(this, projectChecksum.AdditionalDocuments.Checksums).ConfigureAwait(false);
+            await CollectChecksumChildrenAsync(this, projectChecksum.AnalyzerConfigDocuments.Checksums).ConfigureAwait(false);
 
             await _assetProvider.SynchronizeAssetsAsync(
-                assetHint: projectChecksum.ProjectId, checksums, cancellationToken).ConfigureAwait(false);
+                assetHint: projectChecksum.ProjectId, checksums, results: null, cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Host/ChecksumSynchronizer.cs:104]: same here.

File: src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs [hunk 6]
```diff
@@ -29,39 +29,34 @@ internal sealed partial class AssetProvider(Checksum solutionChecksum, SolutionA
...
         // bulk synchronize checksums first
         var syncer = new ChecksumSynchronizer(this);
-        await syncer.SynchronizeAssetsAsync(assetHint, checksums, cancellationToken).ConfigureAwait(false);
+        await syncer.SynchronizeAssetsAsync(assetHint, checksums, results, cancellationToken).ConfigureAwait(false);
 
-        // this will be fast since we actually synchronized the checksums above.
-        using var _ = ArrayBuilder<ValueTuple<Checksum, T>>.GetInstance(checksums.Count, out var list);
-        foreach (var checksum in checksums)
-            list.Add(ValueTuple.Create(checksum, GetRequiredAsset<T>(checksum)));
+        using var _2 = ArrayBuilder<ValueTuple<Checksum, T>>.GetInstance(checksums.Count, out var list);
```
Comment (ToddGrun) [inline src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs:57]: >using var _2 = ArrayBuilder<ValueTuple<Checksum, T>>.GetInstance(checksums.Count, out var list); [](http://example.com/codeflow?start=8&length=96)

Thoughts on just newing up an array and using ImmutableCollectionsMarshal to get it into an ImmutableArray for return?

File: src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs [hunk 7]
```diff
@@ -136,24 +131,60 @@ public async ValueTask SynchronizeProjectAssetsAsync(ProjectStateChecksums proje
...
+                }
+            }
 
             if (usePool)
                 s_checksumPool.Free(missingChecksums);
         }
+
+        return;
+
+        void AddResult(Checksum checksum, object result)
```
Comment (ToddGrun) [inline src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs:183]: >AddResult [](http://example.com/codeflow?start=13&length=9)

nit: I thought we lower cased nested function names

**Review by ToddGrun — APPROVED** [review]
:shipit:

File: src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs]: done.

**Pull request merged**